packages/eslint-plugin/src/rules/prefer-destructuring.ts

```diff
@@ -0,0 +1,232 @@
+      const { program, esTreeNodeToTSNodeMap } = getParserServices(context);
+      const typeChecker = program.getTypeChecker();
+      const baseRules = baseRule.create(context);
+      const rules =
+        leftNode.type === AST_NODE_TYPES.Identifier &&
+        leftNode.typeAnnotation === undefined
+          ? baseRules
+          : baseRule.create(noFixContext(context));
+      if (
+        'typeAnnotation' in leftNode &&
+        leftNode.typeAnnotation !== undefined &&
+        !enforceForDeclarationWithTypeAnnotation
+      ) {
+        return;
+      }
+
+      if (
+        rightNode != null &&
+        isArrayLiteralIntegerIndexAccess(rightNode) &&
+        rightNode.object.type !== AST_NODE_TYPES.Super
+      ) {
+        const tsObj = esTreeNodeToTSNodeMap.get(rightNode.object);
+        const objType = typeChecker.getTypeAtLocation(tsObj);
+        if (!isTypeIterableType(objType, typeChecker)) {
+          if (
+            !enforceForRenamedProperties ||
+            !getNormalizedEnabledType(reportNode.type, 'object')
+          ) {
+            return;
+          }
+          context.report({
+            node: reportNode,
+            messageId: 'preferDestructuring',
+            data: { type: 'object' },
+          });
+          return;
+        }
+      }
+
+      if (reportNode.type === AST_NODE_TYPES.AssignmentExpression) {
+        rules.AssignmentExpression(reportNode);
+      } else {
+        rules.VariableDeclarator(reportNode);
+      }
+    }
+
+    function getNormalizedEnabledType(
+      nodeType:
+        | AST_NODE_TYPES.VariableDeclarator
+        | AST_NODE_TYPES.AssignmentExpression,
+      destructuringType: 'array' | 'object',
+    ): boolean | undefined {
+      if ('object' in enabledTypes || 'array' in enabledTypes) {
+        return enabledTypes[destructuringType];
+      }
+      return enabledTypes[nodeType as keyof typeof enabledTypes][
+        destructuringType as keyof (typeof enabledTypes)[keyof typeof enabledTypes]
+      ];
+    }
+  },
+});
+
+type Context = TSESLint.RuleContext<MessageIds, Options>;
+
+function noFixContext(context: Context): Context {
+  const customContext: {
+    report: Context['report'];
+  } = {
+    report: (descriptor): void => {
+      context.report({
+        ...descriptor,
+        fix: undefined,
+      });
+    },
+  };
+
+  // we can't directly proxy `context` because its `report` property is non-configurable
+  // and non-writable. So we proxy `customContext` and redirect all
+  // property access to the original context except for `report`
+  return new Proxy<Context>(customContext as typeof context, {
+    get(target, path, receiver): unknown {
+      if (path !== 'report') {
+        return Reflect.get(context, path, receiver);
+      }
+      return Reflect.get(target, path, receiver);
+    },
+  });
+}
+
+function isTypeIterableType(
+  type: ts.Type,
+  typeChecker: ts.TypeChecker,
+): boolean {
+  if (isTypeAnyType(type)) {
+    return true;
```

Comment:
[Question] We can't assume that `any`s are or are not iterable. Seeing this return `true` confused me. Let's either return `false` or rename the function to `isTypeAnyOrIterableType`.